test/io/client_test.dart

```diff
@@ -132,4 +146,22 @@ void main() {
 
     expect(socket, isNotNull);
   });
+
+  test('runWithClient', () {
+    late http.Client client;
+    http.runWithClient(() => client = http.Client(), () => TestClient());
```

Comment:
Nit: Can use constructor tear-offs, so `TestClient.new` instead of `() => TestClient()`

Reply:
http works back to dart 2.14 so I can't use tearoffs.